extensions/common/android/xwalk_extension_android.cc

```diff
@@ -162,21 +178,32 @@ XWalkExtensionAndroidInstance::~XWalkExtensionAndroidInstance() {
 void XWalkExtensionAndroidInstance::HandleMessage(
     scoped_ptr<base::Value> msg) {
   std::string value;
-
-  if (!msg->GetAsString(&value)) {
-    return;
-  }
+  const base::BinaryValue* binary_value = nullptr;
 
   JNIEnv* env = base::android::AttachCurrentThread();
-  ScopedJavaLocalRef<jstring> buffer(env, env->NewStringUTF(value.c_str()));
   ScopedJavaLocalRef<jobject> obj = java_ref_.get(env);
   if (obj.is_null()) {
     LOG(ERROR) << "No valid Java object is referenced for message routing";
     return;
   }
 
-  Java_XWalkExtensionAndroid_onMessage(
-      env, obj.obj(), getID(), buffer.obj());
+  if (msg->GetAsString(&value)) {
+    ScopedJavaLocalRef<jstring> buffer(env, env->NewStringUTF(value.c_str()));
+    Java_XWalkExtensionAndroid_onMessage(
+        env, obj.obj(), getID(), buffer.obj());
+  } else if (msg->GetAsBinary(&binary_value)) {
+    ScopedJavaLocalRef<jbyteArray> buffer(
+        env, env->NewByteArray(binary_value->GetSize()));
+    env->SetByteArrayRegion(
+        buffer.obj(), 0, binary_value->GetSize(),
+        reinterpret_cast<jbyte*>(const_cast<char*>(binary_value->GetBuffer())));
+    Java_XWalkExtensionAndroid_onBinaryMessage(
+        env, obj.obj(), getID(), buffer.obj());
+  } else {
+    NOTREACHED() << "Failed to decode message as either string or binary blob";
+  }
+
+  return;
```

Comment:
This `return` here is also unnecessary, you're about to leave the function anyway.